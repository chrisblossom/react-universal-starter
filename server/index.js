/* @flow */

/* eslint-disable import/no-dynamic-require */

import 'babel-polyfill';

import path from 'path';
import express from 'express';
import http from 'http';

import Renderer from './renderer';

const app = express();
const server = http.createServer(app);

const publicPath = path.resolve(__dirname, '../public');
const render = new Renderer();

app.set('etag', false);
app.use('/dist/', express.static(publicPath));


app.use((request, response) => {
    const rendered = render();

    response.send(`<!DOCTYPE html>${rendered}`);
});

const port = process.env.PORT ? parseInt(process.env.PORT, 10) : 3000;

server.listen(port, error => {
    if (error) {
        console.error('express start error', error);

        return;
    }

    console.log('express server started on port', port);
});
