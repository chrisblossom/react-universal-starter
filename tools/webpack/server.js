/* @flow */

/* eslint-disable strict */

'use strict';

const webpack = require('webpack');
const nodeExternals = require('webpack-node-externals');
const config = require('../config/env');
const shared = require('./shared');

const nodeModules = nodeExternals({
    whitelist: [
        /^react-universal-component/,
        /^require-universal-module/,
        /^webpack-flush-chunks/,
    ],
});

const { __DEVELOPMENT__, __PRODUCTION_RELEASE__ } = config.globals;
const { buildPath, srcPath } = config.paths;

const server = {
    name: 'server',
    devtool: __DEVELOPMENT__ ? 'cheap-module-source-map' : 'source-map',
    target: 'node',
    node: {
        console: false,
        global: false,
        process: false,
        Buffer: false,
        __filename: false,
        __dirname: false,
        setImmediate: false,
    },

    externals: [nodeModules],
    context: srcPath,

    entry: {
        main: ['server.js'],
    },

    output: {
        path: buildPath,
        filename: 'server-frontend.js',
        chunkFilename: 'server-frontend-[name]-[chunkhash].js',
        libraryTarget: 'commonjs2',
        publicPath: '/dist/',
    },

    resolve: {
        modules: [srcPath, 'node_modules'],
        extensions: ['.js'],
    },

    module: {
        rules: [
            {
                test: /\.jsx?$|.js.flow$/,
                exclude: /node_modules/,
                use: [
                    {
                        loader: 'babel-loader',
                        options: {
                            cacheDirectory: true,
                            babelrc: false,
                            presets: [
                                [
                                    'env',
                                    {
                                        targets: {
                                            node: '6.10',
                                        },
                                        modules: false,
                                        useBuiltIns: true,
                                    },
                                ],
                                'react',
                            ],
                            plugins: [
                                'syntax-dynamic-import',
                                'transform-object-rest-spread',
                            ],
                        },
                    },
                ],
            },

            {
                test: /\.css$/,
                exclude: /node_modules/,
                use: [
                    {
                        // /locals lets us remove ExtractTextPlugin on the server
                        loader: 'css-loader/locals',
                        options: {
                            sourceMap: false,
                            modules: true,
                            importLoaders: 1,
                            localIdentName: '[local]___[hash:base64:5]',
                        },
                    },
                    {
                        loader: 'postcss-loader',
                        options: {
                            sourceMap: false,
                        },
                    },
                ],
            },

            {
                test: /\.scss$/,
                use: [
                    {
                        loader: 'css-loader',
                        options: {
                            sourceMap: false,
                            modules: true,
                            importLoaders: 2,
                            localIdentName: '[local]___[hash:base64:5]',
                        },
                    },
                    {
                        loader: 'postcss-loader',
                        options: {
                            sourceMap: false,
                        },
                    },
                    {
                        loader: 'sass-loader',
                        options: {
                            sourceMap: false,
                            outputStyle: 'expanded',
                            sourceMapContents: false,
                        },
                    },
                ],
            },
        ],
    },

    plugins: [
        new webpack.BannerPlugin({
            banner: `require('source-map-support').install();`,
            raw: true,
            entryOnly: false,
        }),

        new webpack.optimize.LimitChunkCountPlugin({
            maxChunks: 1,
        }),

        new webpack.DefinePlugin(
            Object.assign({}, shared.definePlugin, {
                __CLIENT__: false,
                __SERVER__: true,
            })
        ),
    ],
};

if (__DEVELOPMENT__ === true) {
    server.plugins.unshift(new webpack.NamedModulesPlugin());
}

if (__PRODUCTION_RELEASE__ === false) {
    server.plugins.unshift(new webpack.NoEmitOnErrorsPlugin());
}

if (__DEVELOPMENT__ === false) {
    server.plugins.unshift(new webpack.HashedModuleIdsPlugin());
}

module.exports = server;
