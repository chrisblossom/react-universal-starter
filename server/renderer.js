/* @flow */

import path from 'path';
import { readFileSync } from 'fs';
import { createHash } from 'crypto';
import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

const rootPath = process.cwd();
const publicPath = path.resolve(__dirname, '../public');

function getHash(filePath) {
    const fileBody = readFileSync(filePath);
    return createHash('sha1').update(fileBody).digest('hex');
}

function Renderer() {
    const serverEntryFile = path.resolve(rootPath, 'build/server-frontend.js');
    const webpackStatsFile = path.resolve(rootPath, 'build/webpack-stats.json');

    let Html = require(serverEntryFile).default;
    let webpackStats = require(webpackStatsFile);
    let previousWebpackStatsFileHash;
    let previousServerEntryFileHash;

    return function render() {
        if (__PRODUCTION_RELEASE__ === false) {
            const webpackStatsFileHash = getHash(webpackStatsFile);
            const serverEntryFileHash = getHash(serverEntryFile);

            if (
                webpackStatsFileHash !== previousWebpackStatsFileHash &&
                serverEntryFileHash !== previousServerEntryFileHash
            ) {
                previousWebpackStatsFileHash = webpackStatsFileHash;
                previousServerEntryFileHash = serverEntryFileHash;

                delete require.cache[require.resolve(serverEntryFile)];
                delete require.cache[require.resolve(webpackStatsFile)];

                Html = require(serverEntryFile).default;
                webpackStats = require(webpackStatsFile);
            }
        }

        return renderToStaticMarkup(
            <Html clientStats={webpackStats} outputPath={publicPath} />,
        );
    };
}

export default Renderer;
