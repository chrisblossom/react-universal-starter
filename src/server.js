/* @flow */

import React from 'react';
import { renderToStaticMarkup, renderToString } from 'react-dom/server';
import { existsSync } from 'fs';
import path from 'path';
import { flushWebpackRequireWeakIds } from 'react-loadable';

import flushChunks from 'webpack-flush-chunks';

import App from './app';

const rootPath = path.resolve(__dirname);
const publicPath = path.resolve(rootPath, 'public');

const getScripts = (scripts: Array<string>) => {
    return scripts.reduce(
        (acc, script: string) => {
            const scriptPath = `/dist/${script}`;

            const preload = (
                <link
                    rel="preload"
                    href={scriptPath}
                    key={script}
                    as="script"
                />
            );

            const scriptTag = (
                <script
                    src={scriptPath}
                    key={script}
                    type="text/javascript"
                    defer
                />
            );

            return {
                preload: [...acc.preload, preload],
                scripts: [...acc.scripts, scriptTag],
            };
        },
        {
            preload: [],
            scripts: [],
        },
    );
};

const getStyles = (styles: Array<string>) => {
    return styles.map((style: string) => {
        const stylePath = `/dist/${style}`;

        return (
            <link
                href={stylePath}
                key={style}
                media="screen, projection"
                rel="stylesheet"
                type="text/css"
                charSet="UTF-8"
            />
        );
    });
};

/**
 * Used in development
 */
const vendorDllFileExists = existsSync(
    path.resolve(publicPath, 'vendor.dll.js'),
);

type PropsType = {
    clientStats: Object,
    outputPath: string,
};

function Html({ clientStats, outputPath }: PropsType) {
    const content = renderToString(<App />);

    const moduleIds = flushWebpackRequireWeakIds();

    const assets = flushChunks(clientStats, {
        before: ['bootstrap'],
        after: ['main'],

        moduleIds,

        // only needed if serving css rather than an external stylesheet
        outputPath,
    });

    const { preload, scripts } = getScripts(assets.scripts);
    const styles = getStyles(assets.stylesheets);

    return (
        <html lang="en-US">
            <head>
                <title>react-safe-universal-inputs demo</title>

                {styles}
                {preload}
            </head>

            <body>
                <br />
                <br />
                <br />
                <br />
                <br />
                <div id="root" dangerouslySetInnerHTML={{ __html: content }} />

                {vendorDllFileExists &&
                    <script
                        src={'/dist/vendor.dll.js'}
                        defer
                        charSet="UTF-8"
                    />}

                {scripts}
            </body>
        </html>
    );
}

type Options = {
    clientStats: Object,
    outputPath: string,
};

const html = (options: Options) => {
    return (
        request: express$Request,
        response: express$Response,
        // eslint-disable-next-line no-unused-vars
        next: express$NextFunction,
    ) => {
        const rendered = renderToStaticMarkup(<Html {...options} />);

        response.send(`<!DOCTYPE html>${rendered}`);
    };
};

export { getScripts, getStyles };
export default html;
