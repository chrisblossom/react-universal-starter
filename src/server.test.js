/* @flow */

import { /* html, */ getScripts, getStyles } from './server';

// it('renders html', () => {
// const result = html();
// });

it('getScripts', () => {
    const scripts = ['bootstrap.js', 'main.cb66a042.js'];

    const result = getScripts(scripts);

    expect(result).toMatchSnapshot();
});

it('getStyles', () => {
    const styles = ['main.css', 'second.css'];

    const result = getStyles(styles);

    expect(result).toMatchSnapshot();
});
