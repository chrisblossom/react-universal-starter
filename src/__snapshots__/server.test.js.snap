// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getScripts 1`] = `
Object {
  "preload": Array [
    <link
      as="script"
      href="/dist/bootstrap.js"
      rel="preload"
/>,
    <link
      as="script"
      href="/dist/main.cb66a042.js"
      rel="preload"
/>,
  ],
  "scripts": Array [
    <script
      defer={true}
      src="/dist/bootstrap.js"
      type="text/javascript"
/>,
    <script
      defer={true}
      src="/dist/main.cb66a042.js"
      type="text/javascript"
/>,
  ],
}
`;

exports[`getStyles 1`] = `
Array [
  <link
    charSet="UTF-8"
    href="/dist/main.css"
    media="screen, projection"
    rel="stylesheet"
    type="text/css"
/>,
  <link
    charSet="UTF-8"
    href="/dist/second.css"
    media="screen, projection"
    rel="stylesheet"
    type="text/css"
/>,
]
`;
